File: ios/NativeInputModule.swift
import ExpoModulesCore

public class NativeInputModule: Module {
    public func definition() -> ModuleDefinition {
        
        Name("NativeInput")
        
        View(NativeInputView.self) {
            
            Prop("text") {(view: NativeInputView, text: String) in
                view.textField.text = text
            }
            
            Prop("keyboardType") { (view: NativeInputView, type: String) in
                switch type {
                    case "numeric": view.textField.keyboardType = .numberPad
                    case "phone": view.textField.keyboardType = .phonePad
                    case "email": view.textField.keyboardType = .emailAddress
                    case "password": view.textField.keyboardType = .default
                    default: view.textField.keyboardType = .default
                }
            }
            
            Prop("autoComplete") { (view: NativeInputView, hint: String) in
                if #available(iOS 10.0, *) {
                    switch hint {
                        case "email": view.textField.textContentType = .emailAddress
                        case "password": view.textField.textContentType = .password
                        case "newPassword": view.textField.textContentType = .newPassword
                        case "username": view.textField.textContentType = .username
                        case "phone": view.textField.textContentType = .telephoneNumber
                        default: view.textField.textContentType = .none
                     }
                 }
             }
            
            Prop("returnKeyType") { (view: NativeInputView, keyType: String) in
                switch keyType {
                    case "done": view.textField.returnKeyType = .done
                    case "go": view.textField.returnKeyType = .go
                    case "next": view.textField.returnKeyType = .next
                    case "search": view.textField.returnKeyType = .search
                    case "send": view.textField.returnKeyType = .send
                    default: view.textField.returnKeyType = .default
                 }
            }

            Prop("padding") { (view: NativeInputView, value: [String: Int]?) in
                guard let value = value else { return }
                let top = CGFloat(value["top"] ?? 0)
                let right = CGFloat(value["right"] ?? 0)
                let bottom = CGFloat(value["bottom"] ?? 0)
                let left = CGFloat(value["left"] ?? 0)
                view.setPadding(top, right, bottom, left)
            }

            Prop("borderColors") { (view: NativeInputView, colors: [String: String]) in
                view.currentColors = colors
            }

            Prop("error") { (view: NativeInputView, hasError: Bool) in
                view.isError = hasError
            }

            Prop("disabled") { (view: NativeInputView, isDisabled: Bool) in
                view.isDisabled = isDisabled
            }

            Prop("secureTextEntry") { (view: NativeInputView, isSecure: Bool) in
                view.textField.isSecureTextEntry = isSecure
            }

            Prop("rightIcon") { (view: NativeInputView, source: [String: Any]?) in
                view.setRightIcon(source)
            }
            
            AsyncFunction("setText") { (view: NativeInputView, text: String) in
                view.textField.text = text
            }
            
            AsyncFunction("isFocused") { (view: NativeInputView) in
                view.isFocused
            }

            AsyncFunction("focus") { (view: NativeInputView) in
                view.textField.becomeFirstResponder()
            }

            AsyncFunction("blur") { (view: NativeInputView) in
                view.textField.resignFirstResponder()
            }
        
            Events(
                "onInputChange",
                "onInputSubmit",
                "onInputFocus",
                "onInputBlur",
                "onRightIconClick"
            )
        }
    }
}

File: android/src/main/java/expo/modules/nativeinput/NativeInputModule.kt
package expo.modules.nativeinput

import android.os.Build
import android.text.InputType
import android.view.View
import android.view.inputmethod.EditorInfo
import androidx.autofill.HintConstants
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class NativeInputModule : Module() {

  override fun definition() = ModuleDefinition {
    Name("NativeInput")

    View(NativeInputView::class) {
      Events("onInputFocus", "onInputBlur", "onInputChange", "onInputSubmit", "onRightIconClick")

      Prop("keyboardType") { view: NativeInputView, type: String ->
        val mappedType = when (type) {
          "numeric" -> InputType.TYPE_CLASS_NUMBER
          "phone" -> InputType.TYPE_CLASS_PHONE
          "email" -> InputType.TYPE_TEXT_VARIATION_EMAIL_ADDRESS
          "password" -> InputType.TYPE_TEXT_VARIATION_PASSWORD
          else -> InputType.TYPE_CLASS_TEXT
        }
        view.editText.inputType = mappedType
      }

      Prop("autoComplete") { view: NativeInputView, hint: String ->
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
          val autofillHints = when(hint) {
            "creditCardExpirationDate" -> HintConstants .AUTOFILL_HINT_CREDIT_CARD_EXPIRATION_DATE
            "creditCardExpirationDay" -> HintConstants.AUTOFILL_HINT_CREDIT_CARD_EXPIRATION_DAY
            "creditCardExpirationMonth" -> HintConstants.AUTOFILL_HINT_CREDIT_CARD_EXPIRATION_MONTH
            "creditCardExpirationYear" -> HintConstants.AUTOFILL_HINT_CREDIT_CARD_EXPIRATION_YEAR
            "creditCardNumber" -> HintConstants.AUTOFILL_HINT_CREDIT_CARD_NUMBER
            "creditCardSecurityCode" -> HintConstants.AUTOFILL_HINT_CREDIT_CARD_SECURITY_CODE
            "emailAddress" -> HintConstants.AUTOFILL_HINT_EMAIL_ADDRESS
            "name" -> HintConstants.AUTOFILL_HINT_PERSON_NAME_GIVEN
            "familyName" -> HintConstants.AUTOFILL_HINT_PERSON_NAME_FAMILY
            "password" -> HintConstants.AUTOFILL_HINT_PASSWORD
            "newPassword" -> HintConstants.AUTOFILL_HINT_NEW_PASSWORD
            "phone" -> HintConstants.AUTOFILL_HINT_PHONE_NUMBER
            "postalAddress" -> HintConstants.AUTOFILL_HINT_POSTAL_ADDRESS
            "postalCode" -> HintConstants.AUTOFILL_HINT_POSTAL_CODE
            "username" -> HintConstants.AUTOFILL_HINT_USERNAME
            "newUsername" -> HintConstants.AUTOFILL_HINT_NEW_USERNAME
            else -> ""
          }
          view.editText.setAutofillHints(autofillHints)
        }
      }

      Prop("importantForAutofill") {view: NativeInputView, important: String ->
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
          val importance = when(important) {
            "auto" -> View.IMPORTANT_FOR_AUTOFILL_AUTO
            "yes" -> View.IMPORTANT_FOR_AUTOFILL_YES
            "no" -> View.IMPORTANT_FOR_AUTOFILL_NO
            "yesExcludeDescendants" -> View.IMPORTANT_FOR_AUTOFILL_YES_EXCLUDE_DESCENDANTS
            "noExcludeDescendants" -> View.IMPORTANT_FOR_AUTOFILL_NO_EXCLUDE_DESCENDANTS
            else -> View.IMPORTANT_FOR_AUTOFILL_AUTO
          }
          view.importantForAutofill = importance
        }
      }

      Prop("returnKeyType") {view: NativeInputView, keyType: String ->
        val imeOption = when (keyType) {
          "done" -> EditorInfo.IME_ACTION_DONE
          "go" -> EditorInfo.IME_ACTION_GO
          "next" -> EditorInfo.IME_ACTION_NEXT
          "search" -> EditorInfo.IME_ACTION_SEARCH
          "send" -> EditorInfo.IME_ACTION_SEND
          else -> EditorInfo.IME_ACTION_UNSPECIFIED
        }
        view.editText.isSingleLine = true
        view.editText.imeOptions = imeOption
      }

      Prop("padding") { view: NativeInputView, value: Map<String, Int>? ->
        view.setPadding(value)
      }

      Prop("borderColors") { view: NativeInputView, colors: Map<String, String> ->
        view.currentColors = colors
        view.updateBorder()
      }

      Prop("error") { view: NativeInputView, hasError: Boolean ->
        view.isError = hasError
        view.updateBorder()
      }

      Prop("disabled") { view: NativeInputView, isDisabled: Boolean ->
        view.isDisabled = isDisabled
        view.editText.isEnabled = !isDisabled
      }

      Prop("secureTextEntry") { view: NativeInputView, isSecure: Boolean ->
        view.isSecure = isSecure
        view.toggleSecureTextEntry()
      }

      Prop("rightIcon") {view: NativeInputView, source: Map<String, Any>? ->
        view.setRightIcon(source)
      }

      AsyncFunction("isFocused") { view: NativeInputView, ->
        view.editText.isFocused
      }

      AsyncFunction("focus") { view: NativeInputView, ->
        view.editText.requestFocus()
      }

      AsyncFunction("blur") { view: NativeInputView ->
        view.editText.clearFocus()
      }

      AsyncFunction("setText") { view: NativeInputView, text: String ->
        view.editText.setText(text)
      }

      AsyncFunction("setSelection") {  view: NativeInputView, start: Int, end: Int ->
        view.editText.setSelection(start, end)
      }
    }
  }
}
